OpenRA.Mods.Common/Traits/FindLinkOnDeploy.cs

```diff
@@ -0,0 +1,129 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Finds the closest link host on deploy.")]
+	public class FindLinkOnDeployInfo : ConditionalTraitInfo, Requires<LinkClientManagerInfo>
+	{
+		[FieldLoader.Require]
+		[Desc("Linking type.")]
+		public readonly BitSet<LinkType> LinkType;
+
+		[CursorReference]
+		[Desc("Cursor to display when able to (un)deploy the actor.")]
+		public readonly string DeployCursor = "deploy";
+
+		[CursorReference]
+		[Desc("Cursor to display when unable to (un)deploy the actor.")]
+		public readonly string DeployBlockedCursor = "deploy-blocked";
+
+		public override object Create(ActorInitializer init) { return new FindLinkOnDeploy(init.Self, this); }
+	}
+
+	public class FindLinkOnDeploy : ConditionalTrait<FindLinkOnDeployInfo>, IIssueDeployOrder, IResolveOrder, IOrderVoice, IIssueOrder
+	{
+		readonly LinkClientManager manager;
+		public FindLinkOnDeploy(Actor self, FindLinkOnDeployInfo info)
+			: base(info)
+		{
+			manager = self.Trait<LinkClientManager>();
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get
+			{
+				yield return new DeployOrderTargeter("FindLink", 10,
+					() => !IsTraitDisabled && manager.LinkingPossible(Info.LinkType)
+						? Info.DeployCursor
+						: Info.DeployBlockedCursor);
+			}
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, in Target target, bool queued)
+		{
+			if (order.OrderID == "FindLink")
+				return new Order(order.OrderID, self, queued);
+
+			return null;
+		}
+
+		Order IIssueDeployOrder.IssueDeployOrder(Actor self, bool queued)
+			=> new("FindLink", self, queued);
+
+		bool IIssueDeployOrder.CanIssueDeployOrder(Actor self, bool queued)
+			=> !IsTraitDisabled && manager.LinkingPossible(Info.LinkType, queued);
+
+		void IResolveOrder.ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString == "FindLink")
+			{
+				self.QueueActivity(order.Queued, new FindLink(this, manager));
+				self.ShowTargetLines();
+			}
+		}
+
+		string IOrderVoice.VoicePhraseForOrder(Actor self, Order order)
+		{
+			if (order.OrderString == "FindLink")
+				return manager.Info.Voice;
+
+			return null;
+		}
+	}
+
+	public class FindLink : Activity
```

Comment:
It seems to me that activities in the engine are usually either `public` and in the `Activities` namespace or private in respective trait. Why is this one public (and in the same file), but not nested inside `FindLinkOnDeploy`? Nor inside the `Activities` namespace.